File: docs/invoice_management_api.ts
/**
 * Invoice Management API Endpoints
 * Ready-to-use Express routes for Invoice Management
 * 
 * Copy this file to: src/routes/invoice.routes.ts
 * 
 * Dependencies:
 * npm install express multer
 * npm install -D @types/express @types/multer
 */

import { Router, Request, Response, NextFunction } from 'express';
import multer from 'multer';
import { format } from 'date-fns';
import { InvoiceParserService } from '../services/invoice-parser.service';
import {
  Invoice,
  ListInvoicesParams,
  InvoiceStatus,
  PaymentType,
  Currency,
} from '../types/invoice.types';

// ============================================
// TYPE DEFINITIONS
// ============================================

interface AuthenticatedRequest extends Request {
  user?: {
    email: string;
    id: string;
  };
}

interface ApproveInvoiceBody {
  scheduledDate?: string;
  notes?: string;
}

interface RejectInvoiceBody {
  reason: string;
}

interface UpdateInvoiceBody {
  vendorName?: string;
  invoiceNumber?: string;
  amount?: number;
  currency?: Currency;
  invoiceDate?: string;
  dueDate?: string;
  description?: string;
  category?: string;
  paymentType?: PaymentType;
  scheduledDate?: string;
}

// ============================================
// MIDDLEWARE
// ============================================

/**
 * Authentication middleware - Replace with your auth system
 */
const loginRequired = (req: AuthenticatedRequest, res: Response, next: NextFunction): void => {
  const userEmail = req.headers['x-user-email'] as string || (req as any).session?.user_email;
  
  if (!userEmail) {
    res.status(401).json({ status: 'error', message: 'Authentication required' });
    return;
  }
  
  req.user = { email: userEmail, id: userEmail };
  next();
};

/**
 * Get user email from request
 */
const getUserEmail = (req: AuthenticatedRequest): string => {
  return req.user?.email || 'unknown@example.com';
};

// ============================================
// ROUTER SETUP
// ============================================

const router = Router();
const upload = multer({ storage: multer.memoryStorage() });
const invoiceService = new InvoiceParserService();

// Apply auth middleware to all routes
router.use(loginRequired);

// ============================================
// UPLOAD ENDPOINTS
// ============================================

/**
 * POST /api/invoices/upload
 * Upload and parse a single invoice PDF
 * 
 * Request: multipart/form-data with 'file' field
 * Response: {
 *   status: 'success',
 *   invoiceId: 'INV-2024-001',
 *   extractedData: { vendorName, amount, ... },
 *   confidence: 0.95
 * }
 */
router.post(
  '/upload',
  upload.single('file'),
  async (req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> => {
    try {
      const file = req.file;
      
      if (!file) {
        res.status(400).json({ status: 'error', message: 'No file provided' });
        return;
      }

      if (file.originalname === '') {
        res.status(400).json({ status: 'error', message: 'No file selected' });
        return;
      }

      const allowedTypes = ['application/pdf', 'image/png', 'image/jpeg', 'image/jpg'];
      if (!allowedTypes.includes(file.mimetype)) {
        res.status(400).json({ status: 'error', message: 'Invalid file type. Allowed: PDF, PNG, JPG' });
        return;
      }

      const userEmail = getUserEmail(req);
      const result = await invoiceService.parseInvoice(
        file.buffer,
        file.originalname,
        file.mimetype,
        userEmail
      );

      res.json(result);
    } catch (error) {
      console.error('Invoice upload error:', error);
      res.status(500).json({
        status: 'error',
        message: error instanceof Error ? error.message : 'Upload failed',
      });
    }
  }
);

/**
 * POST /api/invoices/upload/bulk
 * Upload and parse multiple invoice PDFs
 * 
 * Request: multipart/form-data with 'files[]' field (multiple files)
 * Response: {
 *   status: 'success',
 *   total: 5,
 *   processed: 4,
 *   failed: 1,
 *   results: [{ filename, status, invoiceId, vendorName, amount, error }]
 * }
 */
router.post(
  '/upload/bulk',
  upload.array('files[]', 50),
  async (req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> => {
    try {
      const files = req.files as Express.Multer.File[];

      if (!files || files.length === 0) {
        res.status(400).json({ status: 'error', message: 'No files provided' });
        return;
      }

      if (files.every((f) => f.originalname === '')) {
        res.status(400).json({ status: 'error', message: 'No files selected' });
        return;
      }

      const allowedTypes = ['application/pdf', 'image/png', 'image/jpeg', 'image/jpg'];
      const validFiles = files
        .filter((f) => allowedTypes.includes(f.mimetype))
        .map((f) => ({
          buffer: f.buffer,
          filename: f.originalname,
          mimeType: f.mimetype,
        }));

      if (validFiles.length === 0) {
        res.status(400).json({ status: 'error', message: 'No valid files provided' });
        return;
      }

      const userEmail = getUserEmail(req);
      const result = await invoiceService.parseInvoicesBulk(validFiles, userEmail);

      res.json(result);
    } catch (error) {
      console.error('Bulk upload error:', error);
      res.status(500).json({
        status: 'error',
        message: error instanceof Error ? error.message : 'Bulk upload failed',
      });
    }
  }
);

// ============================================
// LIST & GET ENDPOINTS
// ============================================

/**
 * GET /api/invoices
 * Get paginated list of invoices with filters
 * 
 * Query Parameters:
 *   - page: number (default 1)
 *   - limit: number (default 50)
 *   - status: pending|approved|rejected|paid
 *   - paymentType: Wire|ACH|Card|PayPal|Venmo|Crypto
 *   - currency: USD|EUR|GBP|...
 *   - dateFrom: YYYY-MM-DD
 *   - dateTo: YYYY-MM-DD
 *   - search: string (vendor name, invoice number)
 *   - sortBy: createdAt|invoiceDate|amount|vendorName|status
 *   - sortOrder: asc|desc
 */
router.get(
  '/',
  async (req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> => {
    try {
      const userEmail = getUserEmail(req);

      const params: ListInvoicesParams = {
        page: parseInt(req.query.page as string) || 1,
        limit: parseInt(req.query.limit as string) || 50,
        status: req.query.status as InvoiceStatus | undefined,
        paymentType: req.query.paymentType as PaymentType | undefined,
        currency: req.query.currency as Currency | undefined,
        dateFrom: req.query.dateFrom as string | undefined,
        dateTo: req.query.dateTo as string | undefined,
        search: req.query.search as string | undefined,
        sortBy: req.query.sortBy as ListInvoicesParams['sortBy'],
        sortOrder: req.query.sortOrder as 'asc' | 'desc' | undefined,
      };

      const result = await invoiceService.listInvoices(userEmail, params);
      res.json(result);
    } catch (error) {
      console.error('List invoices error:', error);
      res.status(500).json({
        status: 'error',
        message: error instanceof Error ? error.message : 'Failed to list invoices',
      });
    }
  }
);

/**
 * GET /api/invoices/summary
 * Get invoice summary statistics
 * 
 * Response: {
 *   status: 'success',
 *   summary: {
 *     totalPending: 12,
 *     totalDue: 47890,
 *     overdue: 3,
 *     awaitingApproval: 5,
 *     paidThisMonth: 124500,
 *     scheduled: 4
 *   }
 * }
 */
router.get(
  '/summary',
  async (req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> => {
    try {
      const userEmail = getUserEmail(req);
      const result = await invoiceService.getSummary(userEmail);
      res.json(result);
    } catch (error) {
      console.error('Summary error:', error);
      res.status(500).json({
        status: 'error',
        message: error instanceof Error ? error.message : 'Failed to get summary',
      });
    }
  }
);

/**
 * GET /api/invoices/export
 * Export invoices to CSV
 * 
 * Query Parameters: Same filters as list invoices
 * Response: CSV file download
 */
router.get(
  '/export',
  async (req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> => {
    try {
      const userEmail = getUserEmail(req);

      const params: ListInvoicesParams = {
        limit: 10000,
        status: req.query.status as InvoiceStatus | undefined,
        dateFrom: req.query.dateFrom as string | undefined,
        dateTo: req.query.dateTo as string | undefined,
      };

      const { invoices } = await invoiceService.listInvoices(userEmail, params);

      const headers = [
        'Invoice ID',
        'Invoice Number',
        'Vendor',
        'Amount',
        'Currency',
        'Invoice Date',
        'Due Date',
        'Status',
        'Payment Type',
        'Category',
        'Description',
      ];

      const rows = invoices.map((inv) => [
        inv.invoiceId,
        inv.invoiceNumber || '',
        inv.vendorName,
        inv.amount,
        inv.currency,
        inv.invoiceDate || '',
        inv.dueDate || '',
        inv.status,
        inv.paymentType || '',
        inv.category || '',
        (inv.description || '').replace(/,/g, ';'),
      ]);

      const csv = [headers, ...rows].map((row) => row.join(',')).join('\n');
      const filename = `invoices_${format(new Date(), 'yyyyMMdd')}.csv`;

      res.setHeader('Content-Type', 'text/csv');
      res.setHeader('Content-Disposition', `attachment; filename=${filename}`);
      res.send(csv);
    } catch (error) {
      console.error('Export error:', error);
      res.status(500).json({
        status: 'error',
        message: error instanceof Error ? error.message : 'Export failed',
      });
    }
  }
);

/**
 * GET /api/invoices/:invoiceId
 * Get detailed invoice information
 */
router.get(
  '/:invoiceId',
  async (req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> => {
    try {
      const { invoiceId } = req.params;
      const result = await invoiceService.getInvoice(invoiceId);

      if (!result) {
        res.status(404).json({ status: 'error', message: 'Invoice not found' });
        return;
      }

      res.json({ status: 'success', invoice: result });
    } catch (error) {
      console.error('Get invoice error:', error);
      res.status(500).json({
        status: 'error',
        message: error instanceof Error ? error.message : 'Failed to get invoice',
      });
    }
  }
);

// ============================================
// UPDATE ENDPOINTS
// ============================================

/**
 * PUT /api/invoices/:invoiceId
 * Update invoice data
 */
router.put(
  '/:invoiceId',
  async (req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> => {
    try {
      const { invoiceId } = req.params;
      const data: UpdateInvoiceBody = req.body;
      const userEmail = getUserEmail(req);

      const result = await invoiceService.updateInvoice(invoiceId, data, userEmail);
      res.json(result);
    } catch (error) {
      console.error('Update invoice error:', error);
      res.status(500).json({
        status: 'error',
        message: error instanceof Error ? error.message : 'Update failed',
      });
    }
  }
);

// ============================================
// APPROVAL WORKFLOW ENDPOINTS
// ============================================

/**
 * POST /api/invoices/:invoiceId/approve
 * Approve an invoice for payment
 * 
 * Request Body: {
 *   scheduledDate?: 'YYYY-MM-DD',
 *   notes?: 'Approved for payment'
 * }
 */
router.post(
  '/:invoiceId/approve',
  async (req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> => {
    try {
      const { invoiceId } = req.params;
      const { scheduledDate, notes }: ApproveInvoiceBody = req.body || {};
      const userEmail = getUserEmail(req);

      const result = await invoiceService.approveInvoice(
        invoiceId,
        userEmail,
        scheduledDate,
        notes
      );

      res.json(result);
    } catch (error) {
      console.error('Approve invoice error:', error);
      res.status(500).json({
        status: 'error',
        message: error instanceof Error ? error.message : 'Approval failed',
      });
    }
  }
);

/**
 * POST /api/invoices/:invoiceId/reject
 * Reject an invoice
 * 
 * Request Body: {
 *   reason: 'Duplicate invoice' (required)
 * }
 */
router.post(
  '/:invoiceId/reject',
  async (req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> => {
    try {
      const { invoiceId } = req.params;
      const { reason }: RejectInvoiceBody = req.body || {};

      if (!reason) {
        res.status(400).json({ status: 'error', message: 'Rejection reason required' });
        return;
      }

      const userEmail = getUserEmail(req);
      const result = await invoiceService.rejectInvoice(invoiceId, userEmail, reason);

      res.json(result);
    } catch (error) {
      console.error('Reject invoice error:', error);
      res.status(500).json({
        status: 'error',
        message: error instanceof Error ? error.message : 'Rejection failed',
      });
    }
  }
);

// ============================================
// DOWNLOAD ENDPOINT
// ============================================

/**
 * GET /api/invoices/:invoiceId/download
 * Get signed URL to download original invoice PDF
 * 
 * Response: {
 *   status: 'success',
 *   downloadUrl: 'https://storage.googleapis.com/...',
 *   expiresIn: 3600
 * }
 */
router.get(
  '/:invoiceId/download',
  async (req: AuthenticatedRequest, res: Response, next: NextFunction): Promise<void> => {
    try {
      const { invoiceId } = req.params;
      const url = await invoiceService.getDownloadUrl(invoiceId);

      if (!url) {
        res.status(404).json({ status: 'error', message: 'PDF not found' });
        return;
      }

      res.json({
        status: 'success',
        downloadUrl: url,
        expiresIn: 3600,
      });
    } catch (error) {
      console.error('Download error:', error);
      res.status(500).json({
        status: 'error',
        message: error instanceof Error ? error.message : 'Download failed',
      });
    }
  }
);

// ============================================
// EXPORT
// ============================================

export default router;

/**
 * Register routes with Express app
 * 
 * Usage:
 * import invoiceRoutes from './routes/invoice.routes';
 * app.use('/api/invoices', invoiceRoutes);
 */
export { router as invoiceRouter };

File: docs/invoice_management_api.py
"""
Invoice Management API Endpoints
Ready-to-use Flask routes for Invoice Management

Copy these endpoints to your main Flask app file.
"""

import os
import json
import csv
import io
import logging
from datetime import datetime
from functools import wraps
from flask import Flask, request, jsonify, send_file, session

logger = logging.getLogger(__name__)


def login_required(f):
    """Decorator - replace with your auth system"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        if not session.get('user_email'):
            return jsonify({'error': 'Authentication required'}), 401
        return f(*args, **kwargs)
    return decorated_function


def register_invoice_routes(app):
    """Register all invoice management routes with the Flask app"""
    
    @app.route('/api/invoices/upload', methods=['POST'])
    @login_required
    def upload_invoice():
        """
        Upload and parse a single invoice PDF
        
        Request: multipart/form-data with 'file' field
        Returns: Extracted invoice data with invoice_id
        """
        if 'file' not in request.files:
            return jsonify({'status': 'error', 'message': 'No file provided'}), 400
        
        file = request.files['file']
        if file.filename == '':
            return jsonify({'status': 'error', 'message': 'No file selected'}), 400
        
        allowed_extensions = {'.pdf', '.png', '.jpg', '.jpeg'}
        ext = os.path.splitext(file.filename)[1].lower()
        if ext not in allowed_extensions:
            return jsonify({'status': 'error', 'message': 'Invalid file type'}), 400
        
        try:
            from services.invoice_parser_service import InvoiceParserService
            parser = InvoiceParserService()
            
            file_content = file.read()
            user_email = session.get('user_email', 'unknown@example.com')
            
            result = parser.parse_invoice(file_content, file.filename, user_email)
            return jsonify(result)
            
        except Exception as e:
            logger.error(f"Invoice upload error: {e}")
            return jsonify({'status': 'error', 'message': str(e)}), 500

    @app.route('/api/invoices/upload/bulk', methods=['POST'])
    @login_required
    def upload_invoices_bulk():
        """
        Upload and parse multiple invoice PDFs
        
        Request: multipart/form-data with 'files[]' field (multiple files)
        Returns: Summary with results for each file
        """
        if 'files[]' not in request.files:
            return jsonify({'status': 'error', 'message': 'No files provided'}), 400
        
        files = request.files.getlist('files[]')
        if not files or all(f.filename == '' for f in files):
            return jsonify({'status': 'error', 'message': 'No files selected'}), 400
        
        allowed_extensions = {'.pdf', '.png', '.jpg', '.jpeg'}
        
        try:
            from services.invoice_parser_service import InvoiceParserService
            parser = InvoiceParserService()
            user_email = session.get('user_email', 'unknown@example.com')
            
            file_list = []
            for file in files:
                ext = os.path.splitext(file.filename)[1].lower()
                if ext in allowed_extensions:
                    file_list.append((file.read(), file.filename))
            
            if not file_list:
                return jsonify({'status': 'error', 'message': 'No valid files'}), 400
            
            result = parser.parse_invoices_bulk(file_list, user_email)
            return jsonify(result)
            
        except Exception as e:
            logger.error(f"Bulk upload error: {e}")
            return jsonify({'status': 'error', 'message': str(e)}), 500

    @app.route('/api/invoices', methods=['GET'])
    @login_required
    def list_invoices():
        """
        Get paginated list of invoices with filters
        
        Query params: page, limit, status, payment_type, currency, 
                      date_from, date_to, search, sort_by, sort_order
        """
        try:
            from services.invoice_parser_service import InvoiceParserService
            parser = InvoiceParserService()
            
            page = request.args.get('page', 1, type=int)
            limit = request.args.get('limit', 50, type=int)
            status = request.args.get('status')
            payment_type = request.args.get('payment_type')
            currency = request.args.get('currency')
            date_from = request.args.get('date_from')
            date_to = request.args.get('date_to')
            search = request.args.get('search', '')
            sort_by = request.args.get('sort_by', 'created_at')
            sort_order = request.args.get('sort_order', 'desc')
            
            user_email = session.get('user_email')
            
            result = parser.list_invoices(
                user_email=user_email,
                page=page,
                limit=limit,
                status=status,
                payment_type=payment_type,
                currency=currency,
                date_from=date_from,
                date_to=date_to,
                search=search,
                sort_by=sort_by,
                sort_order=sort_order
            )
            
            return jsonify(result)
            
        except Exception as e:
            logger.error(f"List invoices error: {e}")
            return jsonify({'status': 'error', 'message': str(e)}), 500

    @app.route('/api/invoices/<invoice_id>', methods=['GET'])
    @login_required
    def get_invoice(invoice_id):
        """Get detailed invoice information"""
        try:
            from services.invoice_parser_service import InvoiceParserService
            parser = InvoiceParserService()
            
            result = parser.get_invoice(invoice_id)
            if not result:
                return jsonify({'status': 'error', 'message': 'Invoice not found'}), 404
            
            return jsonify(result)
            
        except Exception as e:
            logger.error(f"Get invoice error: {e}")
            return jsonify({'status': 'error', 'message': str(e)}), 500

    @app.route('/api/invoices/<invoice_id>', methods=['PUT'])
    @login_required
    def update_invoice(invoice_id):
        """Update invoice data"""
        try:
            from services.invoice_parser_service import InvoiceParserService
            parser = InvoiceParserService()
            
            data = request.get_json()
            user_email = session.get('user_email')
            
            result = parser.update_invoice(invoice_id, data, user_email)
            return jsonify(result)
            
        except Exception as e:
            logger.error(f"Update invoice error: {e}")
            return jsonify({'status': 'error', 'message': str(e)}), 500

    @app.route('/api/invoices/<invoice_id>/approve', methods=['POST'])
    @login_required
    def approve_invoice(invoice_id):
        """Approve an invoice for payment"""
        try:
            from services.invoice_parser_service import InvoiceParserService
            parser = InvoiceParserService()
            
            data = request.get_json() or {}
            user_email = session.get('user_email')
            scheduled_date = data.get('scheduled_date')
            notes = data.get('notes')
            
            result = parser.approve_invoice(
                invoice_id, 
                user_email, 
                scheduled_date=scheduled_date,
                notes=notes
            )
            
            return jsonify(result)
            
        except Exception as e:
            logger.error(f"Approve invoice error: {e}")
            return jsonify({'status': 'error', 'message': str(e)}), 500

    @app.route('/api/invoices/<invoice_id>/reject', methods=['POST'])
    @login_required
    def reject_invoice(invoice_id):
        """Reject an invoice"""
        try:
            from services.invoice_parser_service import InvoiceParserService
            parser = InvoiceParserService()
            
            data = request.get_json()
            if not data or not data.get('reason'):
                return jsonify({'status': 'error', 'message': 'Rejection reason required'}), 400
            
            user_email = session.get('user_email')
            reason = data.get('reason')
            
            result = parser.reject_invoice(invoice_id, user_email, reason)
            return jsonify(result)
            
        except Exception as e:
            logger.error(f"Reject invoice error: {e}")
            return jsonify({'status': 'error', 'message': str(e)}), 500

    @app.route('/api/invoices/<invoice_id>/download', methods=['GET'])
    @login_required
    def download_invoice_pdf(invoice_id):
        """Get signed URL to download original invoice PDF"""
        try:
            from services.invoice_parser_service import InvoiceParserService
            parser = InvoiceParserService()
            
            url = parser.get_download_url(invoice_id)
            if not url:
                return jsonify({'status': 'error', 'message': 'PDF not found'}), 404
            
            return jsonify({
                'status': 'success',
                'download_url': url,
                'expires_in': 3600
            })
            
        except Exception as e:
            logger.error(f"Download error: {e}")
            return jsonify({'status': 'error', 'message': str(e)}), 500

    @app.route('/api/invoices/export', methods=['GET'])
    @login_required
    def export_invoices():
        """Export invoices to CSV"""
        try:
            from services.invoice_parser_service import InvoiceParserService
            parser = InvoiceParserService()
            
            user_email = session.get('user_email')
            status = request.args.get('status')
            date_from = request.args.get('date_from')
            date_to = request.args.get('date_to')
            
            invoices = parser.list_invoices(
                user_email=user_email,
                status=status,
                date_from=date_from,
                date_to=date_to,
                limit=10000
            ).get('invoices', [])
            
            output = io.StringIO()
            writer = csv.DictWriter(output, fieldnames=[
                'invoice_id', 'invoice_number', 'vendor_name', 'amount', 
                'currency', 'invoice_date', 'due_date', 'status', 
                'payment_type', 'category', 'description'
            ])
            writer.writeheader()
            
            for inv in invoices:
                writer.writerow({
                    'invoice_id': inv.get('invoice_id'),
                    'invoice_number': inv.get('invoice_number'),
                    'vendor_name': inv.get('vendor_name'),
                    'amount': inv.get('amount'),
                    'currency': inv.get('currency'),
                    'invoice_date': inv.get('invoice_date'),
                    'due_date': inv.get('due_date'),
                    'status': inv.get('status'),
                    'payment_type': inv.get('payment_type'),
                    'category': inv.get('category'),
                    'description': inv.get('description')
                })
            
            output.seek(0)
            return send_file(
                io.BytesIO(output.getvalue().encode()),
                mimetype='text/csv',
                as_attachment=True,
                download_name=f'invoices_{datetime.now().strftime("%Y%m%d")}.csv'
            )
            
        except Exception as e:
            logger.error(f"Export error: {e}")
            return jsonify({'status': 'error', 'message': str(e)}), 500

    @app.route('/api/invoices/summary', methods=['GET'])
    @login_required
    def get_invoice_summary():
        """Get invoice summary statistics"""
        try:
            from services.invoice_parser_service import InvoiceParserService
            parser = InvoiceParserService()
            
            user_email = session.get('user_email')
            result = parser.get_summary(user_email)
            
            return jsonify(result)
            
        except Exception as e:
            logger.error(f"Summary error: {e}")
            return jsonify({'status': 'error', 'message': str(e)}), 500

    return app
